File: perl-example/scripts/import.pl
#!/usr/bin/perl

use strict;
use warnings;

use File::Copy;
use File::Basename;

#
# Exit Codes
#

use constant EXIT_CODE_VALIDATION_ERROR => 1;
use constant EXIT_CODE_TRANSFORMATION_ERROR => 2;
use constant EXIT_CODE_UNEXPECTED_ERROR => 3;

#
# Utilities
#

sub logMessage {
  my $line = $_[0];

  print STDERR "$line\n";
}

sub printWarning {
  my $line = $_[0];

  print STDOUT "$line\n";
}

sub verifyDir {
  my $dir = $_[0];

  if (! -d $dir) {
    logMessage("directory $dir does not exist or is not a directory!");
    exit EXIT_CODE_UNEXPECTED_ERROR;
  }
}

#
# Parameter Handling
#

my ($inputDir, $outputDir) = @ARGV;

if (not defined $inputDir) {
  logMessage("Input directory was not specified");
  exit EXIT_CODE_UNEXPECTED_ERROR;
}
if (not defined $outputDir) {
  logMessage("Output directory was not specified");
  exit EXIT_CODE_UNEXPECTED_ERROR;
}

verifyDir($inputDir);
verifyDir($outputDir);

#
# Script body
#

logMessage("Here's a log message that will be printed to STDERR");

my @inputFiles = <$inputDir/*>;
foreach my $file (@inputFiles) {
  logMessage("Copying $file to $outputDir");
  printWarning("Here's a validation warning about $file that will be printed to STDOUT");
  copy($file, $outputDir . "/" . basename($file));
}

File: go-example/scripts/bin/import.go
package main

import (
	"fmt"
	"handlers/pkg/vdi"
	"os"
	"path/filepath"
)

//
// Utilities
//

func printWarning(message string) {
	fmt.Println(message)
}

func copyFile(from, to string) {
	read, err := os.Open(from)
	if err != nil {
		vdi.LogMessage("unexpected error: " + err.Error())
		os.Exit(vdi.EXIT_CODE_UNEXPECTED_ERROR)
	}
	defer read.Close()

	write, err := os.Create(to)
	if err != nil {
		vdi.LogMessage("unexpected error: " + err.Error())
		os.Exit(vdi.EXIT_CODE_UNEXPECTED_ERROR)
	}
	defer write.Close()

	write.ReadFrom(read)
}

//
// Parameter Handling
//

func main() {
	if len(os.Args) < 3 {
		vdi.LogMessage("called with too few arguments, callers must provide an input and output directory")
		os.Exit(vdi.EXIT_CODE_UNEXPECTED_ERROR)
	}

	inputDirectory := os.Args[1]
	outputDirectory := os.Args[2]

	vdi.VerifyDir(inputDirectory)
	vdi.VerifyDir(outputDirectory)

	runImport(inputDirectory, outputDirectory)
}

//
// Script Body
//

func runImport(inputDir, outputDir string) {
	vdi.LogMessage("Here's a log message that will be printed to STDERR")

	files, err := os.ReadDir(inputDir)
	if err != nil {
		vdi.LogMessage("unexpected error: " + err.Error())
		os.Exit(vdi.EXIT_CODE_UNEXPECTED_ERROR)
	}

	for _, file := range files {
		vdi.LogMessage("Copying " + file.Name() + " to " + outputDir)
		printWarning("Here's a validation warning about " + file.Name() + " that will be printed to STDOUT")
		copyFile(filepath.Join(inputDir, file.Name()), filepath.Join(outputDir, file.Name()))
	}
}
